### PR TITLE
break on 'this' type in hasConstraintReferenceTo

### PR DESCRIPTION
**Pull request by vladima**
fix proposal for #6592, //cc @ahejlsberg

### CODE REVIEW

**Comment by vladima** [comment]
CI build failed because of some random lint errors in files that were not even affected by this PR, looking into this:
```
src/compiler/types.ts[2497, 14]: missing semicolon
src/compiler/types.ts[2507, 14]: missing semicolon
src/compiler/types.ts[2498, 19]: missing whitespace
src/compiler/types.ts[2498, 20]: missing whitespace
src/compiler/types.ts[2499, 18]: missing whitespace
src/compiler/types.ts[2499, 19]: missing whitespace
src/compiler/types.ts[2500, 20]: missing whitespace
src/compiler/types.ts[2500, 21]: missing whitespace
src/compiler/types.ts[2501, 18]: missing whitespace
src/compiler/types.ts[2501, 19]: missing whitespace
src/compiler/types.ts[2502, 21]: missing whitespace
src/compiler/types.ts[2502, 22]: missing whitespace
```

**Comment by vladima** [comment]
pinging @JKillian: it seems that `tslint@next` has a peer dependency `>=1.7.3 || >=1.7.0-dev.20151003 || >=1.8.0-dev` which no longer works after version of TypeScript nightly builds was bumped up to `1.9...`

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -4285,7 +4285,7 @@ namespace ts {
 
         function hasConstraintReferenceTo(type: Type, target: TypeParameter): boolean {
             let checked: Type[];
-            while (type && type.flags & TypeFlags.TypeParameter && !contains(checked, type)) {
+            while (type && !(type.flags & TypeFlags.ThisType) && type.flags & TypeFlags.TypeParameter && !contains(checked, type)) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: I'd rather you move the new check into the body and `break` and have an explanation.

**Comment by JKillian** [comment]
Thanks for the heads-up @vladima. I'll check to see that things are working with TS 1.9 nightlies and bump that dependency.

**Comment by JKillian** [comment]
@vladima Just published a new `next` release on NPM, hopefully that should fix up the issue here

**Comment by DanielRosenwasser** [comment]
Thanks @JKillian!

**Comment by DanielRosenwasser** [comment]
@vladima I requested a re-run on Travis.

File: src/compiler/checker.ts (hunk 1 above)
Comment (ahejlsberg) [inline src/compiler/checker.ts]: No, the check it fine here with all the other checks.

**Comment by ahejlsberg** [comment]
Fix looks good to me.

**Comment by vladima** [comment]
pinging @mhegazy should it also go to 1.8?

**Comment by mhegazy** [comment]
We will need to port this to 1.8.

**Pull request merged**